### PR TITLE
Identity Deletion Job: Actual Deletion of Identities

### PR DESCRIPTION
**Pull request by Dannyps**
# Readiness checklist

-   [x] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

# Description

The flow is as follows:
1. The `Devices.Jobs.IdentityDeletion` job will execute the ~`UpdateDeletionProcessesCommand`~ `TriggerRipeDeletionProcessesCommand`, which:
    a. Finds all identities which have their DeletionGracePeriod in the past.
    b. marks those as `DeletionStarted`.
    c. returns the relevant IdentityAddresses.
2. foreach returned IdentityAddress:
    a. Sends a notification to it.
    a. Fires a `FindRelationshipsOfIdentityCommand`.
    a. Calls all available `IdentityDeleter.Delete` methods, in no particular order.

## March 1st update
 - IdentityDeletion is deleting all database entries for the given identityAddress considering the data created by current integration tests.

## Pre-merge:
1. We're to condense migrations wherever possible.

### CODE REVIEW

**Review by tnotheis — CHANGES_REQUESTED** [review]
I initially had ~100 comments after my review. I already fixed the small ones, so that the PR doesn't get messed up as much. Maybe you can review my commits to make sure I didn't make any mistakes.

File: Modules/Devices/src/Devices.Jobs.IdentityDeletion/Devices.Jobs.IdentityDeletion.csproj
Comment (NikolaDmitrasinovic) [reply Modules/Devices/src/Devices.Jobs.IdentityDeletion/Devices.Jobs.IdentityDeletion.csproj:1]: Just a note: I'm working on a new branch that has Job on top level and is somewhat tied to this, so this can be resolved when that task is done and merged.

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/UpdateDeletionProcesses/HandlerTests.cs
Comment (Dannyps) [reply Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/UpdateDeletionProcesses/HandlerTests.cs:39]: How's the solution I pushed?

**Review by Dannyps — COMMENTED** [review]
As promised, here are a few comments/commits. I'll continue working on the remaining comments in the meantime.

PS: I also closed some comments.

File: Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231122142533_CascadeIdentityDeletion.cs
Comment (tnotheis) [reply Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231122142533_CascadeIdentityDeletion.cs:35]: Okay

File: Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231208002217_DeletionProcess.cs
Comment (tnotheis) [reply Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231208002217_DeletionProcess.cs:11]: Okay, seems like I got confused there. Maybe it's because there is only one migration for SqlServer, but two for Postgres. Why is that?

File: Modules/Devices/src/Devices.Application/Infrastructure/PushNotifications/DeletionProcess/DeletionStartsNotification.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Infrastructure/PushNotifications/DeletionProcess/DeletionStartsNotification.cs:8]: Why not? The config is a public static class. So you can use `IdentityDeletionConfiguration.DeletionStartsNotification.Text`.

**Review by Dannyps — COMMENTED** [review]
Here's another batch :)

File: Modules/Relationships/src/Relationships.Infrastructure.Database.Postgres/Migrations/20231220094204_UpdateDiscriminator.cs
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure.Database.Postgres/Migrations/20231220094204_UpdateDiscriminator.cs:1]: Okay, thank you for the clarification.

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/UpdateDeletionProcesses/HandlerTests.cs
Comment (tnotheis) [reply Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/UpdateDeletionProcesses/HandlerTests.cs]: I'm not 100% convinced, but I think I can live with it. :)

File: Modules/Devices/test/Devices.Domain.Tests/Identities/DeletionProcessTests.cs
Comment (tnotheis) [reply Modules/Devices/test/Devices.Domain.Tests/Identities/DeletionProcessTests.cs]: It's because the tests should use the same interface as the real users of the code. And according to the DDD practices you shouldn't use Entities directly, but only via their Aggregates.

File: Modules/Devices/src/Devices.Jobs.IdentityDeletion/Devices.Jobs.IdentityDeletion.csproj
Comment (tnotheis) [reply Modules/Devices/src/Devices.Jobs.IdentityDeletion/Devices.Jobs.IdentityDeletion.csproj:1]: You can see the intended location here (this is in the `nmshdb-39-automatic-cancellation-of-a-deletion-process` branch): 
https://github.com/nmshd/backbone/tree/nmshdb-39-automatic-cancellation-of-a-deletion-process/Jobs

File: Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231208002217_DeletionProcess.cs
Comment (tnotheis) [reply Modules/Synchronization/src/Synchronization.Infrastructure.Database.Postgres/Migrations/20231208002217_DeletionProcess.cs:11]: Yeah, I agree

**Comment by Dannyps** [comment]
🎉The IdentityDeletion Job is deleting all database entries for the given identityAddresses considering the data created by current integration tests.

File: BuildingBlocks/src/UnitTestTools/Data/TestDataGenerator.cs
Comment (Dannyps) [reply BuildingBlocks/src/UnitTestTools/Data/TestDataGenerator.cs:5]: d4655aa2cc2309ca9f17249f5211316fb0f7d6cb

you did, it seems..?

File: Jobs/src/Devices.Jobs.IdentityDeletion/IdentityDeletionJobConfiguration.cs
Comment (Dannyps) [reply Jobs/src/Devices.Jobs.IdentityDeletion/IdentityDeletionJobConfiguration.cs:1]: > => it should be the same job with two different workers.

I believe this can only be handled when we merge this or the referred branch.

File: Jobs/test/Devices.Jobs.IdentityDeletion.Tests/Tests/WorkerTests.cs
Comment (Dannyps) [reply Jobs/test/Devices.Jobs.IdentityDeletion.Tests/Tests/WorkerTests.cs]: I think the old name was good enough 😅
https://github.com/nmshd/backbone/commit/d4655aa2cc2309ca9f17249f5211316fb0f7d6cb#diff-54d6ffa95b7807568f7d1cffa28cb4944cd7fe7e79863f0f44ae0c275fcc1260R21

File: Jobs/test/Devices.Jobs.IdentityDeletion.Tests/Tests/WorkerTests.cs
Comment (Dannyps) [reply Jobs/test/Devices.Jobs.IdentityDeletion.Tests/Tests/WorkerTests.cs]: I will propose a new test, then.

File: Modules/Devices/src/Devices.Application/Identities/Commands/TriggerRipeDeletionProcesses/Handler.cs
Comment (Dannyps) [reply Modules/Devices/src/Devices.Application/Identities/Commands/TriggerRipeDeletionProcesses/Handler.cs]: Please carefully analyse 9bb337b3cf1c3b76941f49d517eefe7688db7061 as a proposal to address the highlighted issue.

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Domain/DomainErrors.cs
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/DomainErrors.cs:54]: I cannot find such an error. Perhaps you meant `DeletionProcessMustBeInStatus`?

Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/DomainErrors.cs:54]: The thing is that that particular error is thrown when a DeletionProcess is not is one of two statuses, which does not seem to be the scenario that that `DeletionProcessMustBeInStatus`was created for.

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:210]: This check is needed here because of the assignment in the following line. We could move it to the DeletionProperty domain class, but there is no navigation property to the Identity there. I'm not sure how we could work around this issue.

File: Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionProcess.cs
Comment (Dannyps) [reply Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionProcess.cs:128]: Please check 4e9bbe08938de5536bac5c6da23eea6c5b237b9b for a suggestion of implementation

File: Modules/Devices/src/Devices.Domain/DomainErrors.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/DomainErrors.cs]: Why do you even check for `IsActive`? The only status you're interested in is `Approved`. If it is still waiting for approval, you should never ever delete anyway.

Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/DomainErrors.cs]: Also: I am pretty sure that I commented this in one of @NikolaVetnic's PRs, but: the error code in line 51 is dynamic (it contains the status), which is really bad. Could you please fix this? Rephrasing is probably necessary, as just removing the dynamic part results in a kind of "incomplete" English.

File: Modules/Devices/src/Devices.Domain/DomainErrors.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/DomainErrors.cs:56]: > it has not been approved or is not past due approval

"it" gramatically refers to the identity, which doesn't make sense, because an identity can neither be approved, not can it be past due approval

File: Modules/Relationships/src/Relationships.Application/Identities/IdentityDeleter.cs
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Application/Identities/IdentityDeleter.cs:21]: No, because an allocation is created for each peer template fetched by the identity to be deleted. 

But: this comment is kind of outdated anyway. As discussed in another channel, we shouldn't DELETE allocations, but instead replace the address of the identity to be deleted with a placeholder.

File: Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionProcess.cs
Comment (tnotheis) [reply Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionProcess.cs]: That's why I said you can make it private.

File: Modules/Files/test/Files.Infrastructure.Tests/Tests/Repositories/FilesRepositoryTests.cs
Comment (tnotheis) [reply Modules/Files/test/Files.Infrastructure.Tests/Tests/Repositories/FilesRepositoryTests.cs:15]: But you do test that `A.CallTo(() => mockBlobStorage.Remove(A<string>._, file.Id)).MustHaveHappenedOnceExactly()`. So why not test the opposite as well?

Btw, `A<string>._` should be replaced with the address of the identity to be deleted.

Comment (Dannyps) [reply Modules/Files/test/Files.Infrastructure.Tests/Tests/Repositories/FilesRepositoryTests.cs:15]: Please see bd005ba0feb46119741a98e699a83dc6b8beb670

> Btw, A<string>._ should be replaced with the address of the identity to be deleted.

But as far as I can see, that's not the folder (which is the first param in the `Remove` method. It is actually `RootFolder`. See line 42.

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/EntityTypeConfigurations/RelationshipTemplateEntityTypeConfiguration.cs
Comment (Dannyps) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/EntityTypeConfigurations/RelationshipTemplateEntityTypeConfiguration.cs:26]: So, should I remove the Anonymization logic introduced by the previous commit?

Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/EntityTypeConfigurations/RelationshipTemplateEntityTypeConfiguration.cs:26]: No. The Cascade is for templates created by the **deleted** identity. The anonymization is for templates **allocated** by the deleted identity.

File: .github/CODEOWNERS
Comment (tnotheis) [inline .github/CODEOWNERS:1]: What did you do here? Was this your markdown formatter?

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:1]: There is a null warning in line 160

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20240403115614_ActualDeletion.cs
Comment (tnotheis) [inline Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20240403115614_ActualDeletion.cs:26]: Seems like you forgot to recreate the migration after you made the property non-nullable.

File: Modules/Messages/src/Messages.Application/Infrastructure/Persistence/Repository/IMessagesRepository.cs
Comment (tnotheis) [inline Modules/Messages/src/Messages.Application/Infrastructure/Persistence/Repository/IMessagesRepository.cs:13]: Weird formatting

File: Modules/Relationships/src/Relationships.Infrastructure.Database.Postgres/Migrations/20240403115757_ActualDeletion.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure.Database.Postgres/Migrations/20240403115757_ActualDeletion.cs:1]: Why the empty migration?

File: Modules/Relationships/src/Relationships.Infrastructure.Database.SqlServer/Migrations/20240403115724_ActualDeletion.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure.Database.SqlServer/Migrations/20240403115724_ActualDeletion.cs:1]: Why the empty migration?

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS
Comment (tnotheis) [reply .github/CODEOWNERS:1]: Just undo your changes to this file from this branch

Comment (tnotheis) [reply .github/CODEOWNERS:1]: The `*` is a glob pattern

File: Jobs/src/Job.IdentityDeletion/IdentityDeletionJobConfiguration.cs
Comment (Dannyps) [reply Jobs/src/Job.IdentityDeletion/IdentityDeletionJobConfiguration.cs:11]: In that case, shouldn't we provide a default value? (`string.Empty`)

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20240403115614_ActualDeletion.cs
Comment (Dannyps) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20240403115614_ActualDeletion.cs:26]: But the property is marked as non-nullable 🤔

Comment (tnotheis) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/20240403115614_ActualDeletion.cs:26]: Hm, yeah, idk what' I've seen there. 😆

**Review by tnotheis — APPROVED** [review]

**Review by tnotheis — APPROVED** [review]

**Pull request merged**
